tests/manage/z_cluster/test_validate_ceph_config_values.py

```diff
@@ -0,0 +1,50 @@
+import logging
+import pytest
+import collections
+
+
+from ocs_ci.framework.testlib import ManageTest, tier1
+from ocs_ci.framework import config
+from ocs_ci.ocs.ocp import OCP
+from ocs_ci.ocs import constants, defaults
+
+
+log = logging.getLogger(__name__)
+
+
+@tier1
+@pytest.mark.skipif(
+    config.DEPLOYMENT["ceph_debug"],
+    reason="Ceph was configured with customized values by ocs-ci so there is point in validating its config values",
+)
+class TestValidateCephConfigValues(ManageTest):
+    """
+    Test class for Ceph config values validation
+    """
+
+    def test_validate_ceph_config_values(self):
+        """
+        Test case for comparing the cluster's config values of
+        Ceph with the static set of configuration saved in ocs-ci
+
+        """
+        cm_obj = OCP(
+            kind="configmap",
+            namespace=defaults.ROOK_CLUSTER_NAMESPACE,
+            resource_name=constants.ROOK_CONFIG_OVERRIDE_CONFIGMAP,
+        )
+        config_data = cm_obj.get().get("data").get("config")
+        config_data = config_data.split("\n")[1:-1]
+        log.info(
+            "Validating that the Ceph values, configured by ceph-config-override "
+            "confiMap, match the ones stored in ocs-ci"
+        )
+
+        assert collections.Counter(config_data) == collections.Counter(
```

Comment:
It would be better to compare this via `ConfigParser` objects, either comparing these objects or using these objects to serialize canonical string representation to compare as a string.

Reply:
Or we can compare the configs as a strings without parsing, but that would fail in case of whitespace or ordering changes.

Reply:
but what if the key and values are the same but the ordering is different between the 2 configs?

Reply:
> but what if the key and values are the same but the ordering is different between the 2 configs?

My 1st note https://github.com/red-hat-storage/ocs-ci/pull/3908#discussion_r596745971 is discussing exactly that. Order in ini file doesn't change it's semantics, so comparing `ConfigParser` objects handles that.

See this example, two files with different ordering, but the same values:

```
$ cat one.ini 
[foobar]
a = 1
b = 2
$ cat two.ini
[foobar]
b = 2
a = 1
```

```
$ cat test_cp.py 
# -*- coding: utf8 -*-

from configparser import ConfigParser


def test_different_order():
    c1 = ConfigParser()
    c1.read_file(open('one.ini'))
    c2 = ConfigParser()
    c2.read_file(open('two.ini'))
    assert c1 == c2


def test_different_content():
    c2 = ConfigParser()
    c2.read_file(open('two.ini'))
    ca = ConfigParser()
    ca.read_file(open('another.ini'))
    assert c2 != ca
```

```
$ python3 -m pytest test_cp.py -v
============================ test session starts =============================
platform linux -- Python 3.9.2, pytest-6.0.2, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/martin/tmp
plugins: pyfakefs-3.5.8
collected 2 items                                                            

test_cp.py::test_different_order PASSED                                [ 50%]
test_cp.py::test_different_content PASSED                              [100%]

============================= 2 passed in 0.01s ==============================
```

But a negative case shows that the error produced would not be usable:

```
>       assert c2 == ca
E       assert <configparser...x7fd9e4af1fd0> == <configparser...x7fd9e4af1d60>
E         +<configparser.ConfigParser object at 0x7fd9e4af1fd0>
E         -<configparser.ConfigParser object at 0x7fd9e4af1d60>
E         Full diff:
E         - <configparser.ConfigParser object at 0x7fd9e4af1d60>
E         ?                                                  -
E         + <configparser.ConfigParser object at 0x7fd9e4af1fd0>
E         ?  
```

Which means we would need to compare (canonical) string representation generated by ConfigParser, but looking at it I see that ConfigParser now preserves order by default and it can't be easily disabled to generate canonical string export:

```
    def test_different_order():
        c1 = ConfigParser()
        c1.read_file(open('one.ini'))
        c1_str = StringIO()
        c1.write(c1_str)
        c2 = ConfigParser()
        c2.read_file(open('two.ini'))
        c2_str = StringIO()
        c2.write(c2_str)

>       assert c1_str.getvalue() == c2_str.getvalue()
E       AssertionError: assert '[foobar]\na = 1\nb = 2\n\n' == '[foobar]\nb = 2\na = 1\n\n'
E           [foobar]
E         + a = 1
E           b = 2
E         - a = 1
```

Which is pity, as pytest diff feature is able to provide pretty good explanation of the issue.

Ok.